### PR TITLE
New ImageBlock for StreamField

### PR DESCRIPTION
**Pull request by Chiemezuo**
This is an experimental feature I'm demoing for my GSoC proposal.
It is a new StreamField block called `ImageBlock`.
It's a work in progress, but I'd like it to form a basis for my potential conversations about the 'alt text project' with mentors @Stormheg @SaptakS and @thibaudcolas  

This is a sample of what the `ImageBlock` would look like:
image

And this is what a sample of what it would look like from the console output:
image

### CODE REVIEW

**Review by Stormheg — COMMENTED** [review]
Have yet to try this out on a Wagtail installation, but this is already looks amazing to me! Good work 👍

**Comment by Chiemezuo** [comment]
Thank you, @Stormheg :)

File: wagtail/images/models.py
Comment (Stormheg) [inline wagtail/images/models.py:628]: 👍 👏

File: wagtail/images/tests/test_blocks.py
Comment (Chiemezuo) [reply wagtail/images/tests/test_blocks.py:121]: Thanks a bunch for this tip!

File: client/src/entrypoints/images/image-chooser-modified.js
Comment (Stormheg) [reply client/src/entrypoints/images/image-chooser-modified.js:14]: We'll keep it for now and change it the use your Stimulus code once #11202 gets merged @lb- . Hopefully soon! 🤞

Comment (lb-) [reply client/src/entrypoints/images/image-chooser-modified.js:14]: Sounds good.

**Comment by lb-** [comment]
@Chiemezuo  do you mind changing this to a draft PR while it's still a WIP?

**Comment by Chiemezuo** [comment]
Sure, I'll do this. 😅 It was ready for review some days ago, but something did come up again. I'll change it back to a draft.

**Comment by lb-** [comment]
My apologies, I should have read all the comments. If it's close to review then, maybe just update the title to not have the 'work in progress'.

**Comment by Chiemezuo** [comment]
That's completely okay. The comments won't really point out the cause. It was a modification my mentors and I agreed to make. Regardless, it'll for sure be completed when I add the upgrade considerations @lb-

File: client/src/entrypoints/images/image-chooser-modified.js
Comment (thibaudcolas) [reply client/src/entrypoints/images/image-chooser-modified.js:14]: @Chiemezuo can you add a code comment in here to explain this implementation will need rework once #11202 is merged?

File: client/src/entrypoints/images/image-chooser-modified.js
Comment (thibaudcolas) [inline client/src/entrypoints/images/image-chooser-modified.js:13]: For later – for the rich text implementation, we also add an asterisk next to the alt text field label to convey it’s required. Though this is broken currently.

We should be consistent, either remove the currently-broken rich text code that does this, or implement the pattern in both places.

File: wagtail/images/blocks.py
Comment (thibaudcolas) [inline wagtail/images/blocks.py:121]: I’m surprised this isn’t calling the `clean` method from StructBlock? `result = super().clean(value)`

File: wagtail/images/blocks.py
Comment (thibaudcolas) [inline wagtail/images/blocks.py:61]: I think this is worth a docstring to explain to Wagtail users and maintainers that this StructBlock has magic in it so its `value` is an image model instance.

File: wagtail/images/blocks.py
Comment (thibaudcolas) [inline wagtail/images/blocks.py:64]:
```suggestion
        default=False, required=False, label=_("Image is decorative")
```

**Review by thibaudcolas — COMMENTED** [review]
Lots of progress here, there are changes to make but I think this is ready for further review!

Global changes to make:

- Make sure all new text labels (e.g. error messages) are translate-able
- For text files, make sure to `git config --global core.autocrlf true` so the files are with `lf` like endings. To check line endings with git:

```bash
git ls-files --eol <path to files you want to check>
```

Questions:

- What’s our test coverage target for this new code? For example for Telepath adapters?

File: wagtail/images/blocks.py
Comment (Chiemezuo) [reply wagtail/images/blocks.py:64]: I take it I will have to import through `django.utils.translation import gettext as _ `, right?

File: wagtail/images/tests/test_blocks.py
Comment (Chiemezuo) [reply wagtail/images/tests/test_blocks.py:8]: Thank you, for this!

File: wagtail/images/tests/test_blocks.py
Comment (Chiemezuo) [reply wagtail/images/tests/test_blocks.py:89]: Coming right up

File: wagtail/images/blocks.py
Comment (thibaudcolas) [inline wagtail/images/blocks.py:107]: So here as well it seems to me like we need to re-implement. The point of the `bulk_to_python` methods as I understand them is reducing the number of queries by combining DB lookups. Unless we’ve agreed the optimization wasn’t worth chasing?

If it’s not worth chasing – then I’d expect this "bulk" method needs re-implementing to reuse the `to_python` method. Not worth duplicating the code as much as we are currently.

If it is worth chasing, then, seems like we need to use `ImageChooserBlock` / `FieldBlock`’s `bulk_to_python`. Convert our mixed `values` with direct image references and ints to one list of direct image references. Bulk fetch that. Then return with our other StructBlock fields’ data, and appropriate defaults for any image references.

File: wagtail/images/models.py
Comment (thibaudcolas) [inline wagtail/images/models.py:630]: Might be a silly question but this is only done in this specific scenario because… in other scenarios it’s already done elsewhere?

**Review by thibaudcolas — CHANGES_REQUESTED** [review]

File: docs/reference/streamfield/blocks.md
Comment (gasman) [reply docs/reference/streamfield/blocks.md:421]: I can see that the 'is decorative' checkbox is probably redundant in a situation like this, but surely the added value of the contextual alt text field makes it worthwhile?

Realistically I think there'll always be a significant portion of Wagtail devs who aren't particularly switched on to accessibility considerations, and won't be actively thinking "is `ImageBlock` or `ImageChooserBlock` the right choice here?" when putting their models together. For those devs, our goal should be to plant the idea that `ImageBlock` is the right one to reach for by default, even if they're not actively aware of the reasoning _why_ it's the right one. (Indeed, that's a large part of the reason I originally proposed `ImageBlock` as the name rather than, say, `ImageWithAltChooserBlock`...) For that reason, I think there's a good case to make for the documentation erring on the side of using `ImageBlock` more frequently to reinforce that idea, even if the reality is more nuanced.

File: client/src/entrypoints/images/image-chooser-modified.js
Comment (Chiemezuo) [reply client/src/entrypoints/images/image-chooser-modified.js:1]: I think this is a great suggestion!

File: docs/reference/streamfield/blocks.md
Comment (Chiemezuo) [reply docs/reference/streamfield/blocks.md:435]: However, I think the existence of a provision for alt text is a good trade-off, don't you think?

File: wagtail/images/blocks.py
Comment (Chiemezuo) [reply wagtail/images/blocks.py:5]: `gettext_lazy` is a better option

File: wagtail/images/models.py
Comment (Chiemezuo) [reply wagtail/images/models.py:630]: Oh, no. It's not done anywhere else

File: docs/reference/streamfield/blocks.md
Comment (Chiemezuo) [reply docs/reference/streamfield/blocks.md:421]: @Stormheg @SaptakS What are your opinions on this?
I think having alt text is a good trade-off to using the `Image_block` despite the redundancy of the `decorative` checkbox.

Comment (SaptakS) [reply docs/reference/streamfield/blocks.md:421]: I tend to agree with @gasman on this . I think most people might  go for the default option shown in one example, and it's important to tell them that in most cases `ImageBlock` might be the right one. The developers still definitely have the choice to use `ImageChooserBlock` if they want. I agree @thibaudcolas that in this case it's definitely a developer's decision, but I feel more frequently we talk about `ImageBlock` in the examples, the more widely it will get adapted (and also hopefully developers and editors will learn a little more about the different scenarios and cases where they might want it to be decorative or not)

File: wagtail/images/blocks.py
Comment (Chiemezuo) [reply wagtail/images/blocks.py:122]: The problem with that is that using the `super().clean(value)` would fail. Reason being is that the superclass method calls that with the expectation that the value received will be a `dict`, and calls the `.items()` method on it. That will not be the case for us, where the value is supposed to be an Image instance.

File: wagtail/images/blocks.py
Comment (Chiemezuo) [reply wagtail/images/blocks.py:111]: I've made a fix for this. I'll push it shortly.

File: wagtail/images/blocks.py
Comment (Chiemezuo) [reply wagtail/images/blocks.py:70]: I went with the parent class implementation.

File: wagtail/images/blocks.py
Comment (Stormheg) [inline wagtail/images/blocks.py:144]: Is there a way to trigger this code path? I would like to see what this outputs.

If it is rarely executed; would a raw (untranslated) ValueError for the developer be better perhaps? Now we introduce an additional translatable string that is rarely used.

File: docs/tutorial/create_portfolio_page.md
Comment (Chiemezuo) [reply docs/tutorial/create_portfolio_page.md:52]: It's a tricky situation. On the one hand, the caption might remove the need for an alt text, but on the other hand, we are trying to nudge users toward using the `ImageBlock` over the `ImageChooserBlock`. 
Also, there's a chance that the "caption" may be something the editor wants to display, e.g a quote from the person in the image. In contrast, the alt text would be a description of the content in the image itself, specifically for accessibility purposes.

File: wagtail/images/blocks.py
Comment (Chiemezuo) [reply wagtail/images/blocks.py:144]: It's almost never executed. To be frank, I'm not sure how possible it is for an image that isn't an instance of "abstractImage" to be uploaded. Perhaps it's a bit unnecessary and I can remove it. What do you think? Or should I leave it as is?

File: docs/tutorial/create_portfolio_page.md
Comment (thibaudcolas) [reply docs/tutorial/create_portfolio_page.md:52]: I think this is good as-is. The visible caption could be appropriate as alt text but it’s not uncommon for there to be both, when you want the opportunity to provide additional details to screen reader users.

File: wagtail/images/blocks.py
Comment (thibaudcolas) [inline wagtail/images/blocks.py:67]: Seems to me like there might be more work required to override the different methods of StructBlock to be compatible with a value provided as an Image model instance, based on Storm’s feedback here: https://github.com/wagtail/bakerydemo/pull/492#issuecomment-2314665023

For example, the way `get_searchable_content` is implemented, it would be provided with an image instance rather than a struct value. I worry other methods might require a similar treatment (usage counts?).

File: docs/tutorial/create_portfolio_page.md
Comment (SaptakS) [reply docs/tutorial/create_portfolio_page.md:52]: I agree that caption and alt-text can be different so I think this is fine.

File: wagtail/images/blocks.py
Comment (thibaudcolas) [inline wagtail/images/blocks.py:194]: We’re missing an implementation of `get_api_representation`. In my testing, seems like the following would work well. It’s not backwards-compatible, but we couldn’t return the contextual alt text in a backwards-compatible way anyway:

```python
    def get_api_representation(self, value, context=None):
        return self.get_prep_value(value)
```

File: wagtail/images/blocks.py
Comment (thibaudcolas) [inline wagtail/images/blocks.py:65]: I think we’re missing further compatibility work with `StructBlockComparison` (wagtail/admin/compare.py), as currently that code is attempting to access the block’s data as if it was a StructBlock, even though its value is returned as an image.

And we need to implement the block’s `render_basic` to render an image. Most likely something like:

```python
    def render_basic(self, value, context=None):
        return self.child_blocks["image"].render_basic(value)
```

This can all be tested by adding a block to a page, saving a draft, and then going to the "History" view to compare revisions.

**Review by thibaudcolas — CHANGES_REQUESTED** [review]
Based on further reviewing, I think I’ve identified an additional 2-4 code paths that need further work to be compatible with our backwards-compatibility data conversion:

- `get_api_representation`
- `render_basic` and the `get_comparison_class_for_block` logic
- (`get_searchable_content`, already raised)
- Possibly the `map_struct_block_value` in the StreamField migration utilities

What to do for `get_api_representation` and `get_searchable_content` seems straightforward enough. For `get_comparison_class_for_block`, I suspect we need to add explicit support for our new block type there, using `RichTextBlockComparison` as that’s what I believe is in use for ImageChooserBlock. It should work great as long as we also implement `render_basic`.

For `map_struct_block_value`, I have zero experience with this code so not clear what to do (if anything).

File: docs/tutorial/create_portfolio_page.md
Comment (Stormheg) [reply docs/tutorial/create_portfolio_page.md:52]: Alright 👍

File: wagtail/images/blocks.py
Comment (Chiemezuo) [reply wagtail/images/blocks.py:117]: Thanks for this!
I'll look deeper into it.

File: docs/reference/streamfield/blocks.md
Comment (gasman) [inline docs/reference/streamfield/blocks.md:347]: `required=False` is documented here, but is not currently implemented - the block always behaves as `required=True`, because the underlying `ImageChooserBlock` is declared as `required=True` and the `clean` method explicitly rejects an empty value.

**Review by gasman — CHANGES_REQUESTED** [review]

File: wagtail/images/blocks.py
Comment (gasman) [reply wagtail/images/blocks.py:144]: This `clean` method will need wider reworking to support `required=False`.

File: wagtail/images/blocks.py
Comment (gasman) [reply wagtail/images/blocks.py:67]: The missing methods are now (hopefully) all implemented in #12416.

File: wagtail/images/blocks.py
Comment (gasman) [reply wagtail/images/blocks.py:194]: Implemented in https://github.com/wagtail/wagtail/pull/12416/commits/f09fb3d22aa2fa54264ec6e148322b0b92377fe8

File: wagtail/images/blocks.py
Comment (gasman) [reply wagtail/images/blocks.py:65]: Fixed in https://github.com/wagtail/wagtail/pull/12416/commits/09b65b0070299a9077f83f1e9455798cacfa4077

**Comment by gasman** [comment]
> For `map_struct_block_value`, I have zero experience with this code so not clear what to do (if anything).

I don't think anything needs to be done here. The fact that ImageBlock is derived from StructBlock is an internal implementation detail and the child block names are hard-coded, so there is no reason for a developer to be writing migrations that operate on it as a StructBlock (such as moving a child block in or out of it). I've confirmed that a basic rename operation on an ImageBlock works - I guess I should also test moving an ImageBlock around in the stream structure, so I'll try that next.

File: docs/reference/streamfield/blocks.md
Comment (gasman) [reply docs/reference/streamfield/blocks.md:347]: `required=False` now implemented in https://github.com/wagtail/wagtail/pull/12416/commits/01eec77695dad1d0bf7435b62ea1fb83783ff1fa

**Comment by gasman** [comment]
Superseded by #12416

**Pull request closed without merging**
